draft-ietf-tls-esni.md

```diff
@@ -438,58 +425,70 @@ aborts the connection with an "illegal_parameter" alert.
 
 ## Sending an encrypted ClientHello {#send-ech}
 
-In order to send an encrypted ClientHello, the client first determines if it
-supports the server's chosen KEM, as identified by `ECHConfig.kem_id`. If one is
-supported, the client MUST select an appropriate `HpkeCipherSuite` from the list
-of suites offered by the server. If the client does not support the
-corresponding KEM or is unable to select a suitable `HpkeCipherSuite`, it SHOULD
-ignore that `ECHConfig` value and MAY attempt to use another value provided by
-the server. The client MUST NOT send ECH using HPKE algorithms not advertised by
-the server.
-
-Given a compatible `ECHConfig` with fields `public_key` and `kem_id`, carrying
-the `HpkePublicKey` and KEM identifier corresponding to the server, clients
-compute an HPKE encryption context as follows:
+To offer ECH, the client first chooses a suitable ECH configuration. To
+determine if a given `ECHConfig` is suitable, it checks that it supports the KEM
+algorithm identified by `ECHConfig.kem_id` and at least one KDF/AEAD algorithm
+identified by `ECHConfig.cipher_suites`. Once a suitable configuration is found,
+the client selects the cipher suite it will use for encryption. It MUST NOT
+choose a cipher suite not advertised by the configuration.
+
+Next, the client constructs the ClientHelloOuter message just as it does a
+standard ClientHello, with the exception of the following rules:
+
+1. It MUST offer to negotiate TLS 1.3 or above.
+1. It MUST include an "encrypted_client_hello" extension with a payload
+   constructed as described below.
+1. The value of `ECHConfig.public_name` MUST be placed in the "server_name"
+   extension.
+1. It MUST NOT include a PSK or PSK binder. [[OPEN ISSUE: The text needs to be
+   updated to reflect this rule.]]
+
+The client then constructs the ClientHelloInner message just as it does a
+standard ClientHello, with the exception of the following rules:
+
+1. It MUST NOT offer to negotiate TLS 1.2 or below.
+1. It MUST NOT offer to resume any non-ECH PSK or any session for TLS 1.2 and
+   below.
+1. It MAY offer any other extension in the ClientHelloOuter except those that
+   have been incorporated into the ClientHelloInner as described in
+   {{outer-extensions}}.
+1. It MAY copy any other field from the ClientHelloOuter except
+   ClientHelloOuter.random. Instead, It MUST generate a fresh
+   ClientHelloInner.random using a secure random number generator. (See
+   {{flow-client-reaction}}.)
+1. It SHOULD contain TLS padding {{!RFC7685}} as described in {{padding}}.
+
+The client might duplicate non-sensitive extensions in both messages. However,
+implementations need to take care to ensure that sensitive extensions are not
+offered in the CllientHelloOuter. [[OPEN ISSUE: We should provide guidance on
+what extensions are sensitive and suggest suitable substitutes.]]
```

Comment:
Why substitutes? Why not just drop them from ClientHelloOuter entirely?

Reply:
"substitutes" = "innocuous values for sensitive extensions". Is that more clear?